packages/@aws-cdk/core/lib/private/runtime-info.ts

```diff
@@ -1,95 +1,81 @@
- * in the parent directory, not in the ``node_modules`` directory. For this
- * reason, we strip the ``/node_modules`` suffix off each path and use regular
- * module resolution to obtain a reference to ``package.json``.
- *
- * @param fileName a javascript file name.
- * @returns the NPM module infos (aka ``package.json`` contents), or
- *      ``undefined`` if the lookup was unsuccessful.
+ * For a given stack, walks the tree and finds the runtime info for all constructs within the tree.
+ * Returns the unique list of construct info present in the stack,
+ * as long as the construct fully-qualified names match the defined allow list.
  */
-function findNpmPackage(fileName: string): { name: string, version: string, private?: boolean } | undefined {
-  const mod = require.cache[fileName];
-
-  if (!mod?.paths) {
-    // sometimes this can be undefined. for example when querying for .json modules
-    // inside a jest runtime environment.
-    // see https://github.com/aws/aws-cdk/issues/7657
-    // potentially we can remove this if it turns out to be a bug in how jest implemented the 'require' module.
-    return undefined;
+export function constructInfoFromStack(stack: Stack): ConstructInfo[] {
+  function isConstructInfo(value: ConstructInfo | undefined): value is ConstructInfo {
+    return value !== undefined;
   }
 
-  // For any path in ``mod.paths`` that is a node_modules folder, use its parent directory instead.
-  const paths = mod?.paths.map((path: string) => basename(path) === 'node_modules' ? dirname(path) : path);
+  const allConstructInfos = constructsInStack(stack)
+    .map(construct => constructInfoFromConstruct(construct))
+    .filter(isConstructInfo) // Type simplification
+    .filter(info => ALLOWED_FQN_PREFIXES.find(prefix => info.fqn.startsWith(prefix)));
 
-  try {
-    const packagePath = require.resolve(
-      // Resolution behavior changed in node 12.0.0 - https://github.com/nodejs/node/issues/27583
-      nodeMajorVersion >= 12 ? './package.json' : 'package.json',
-      { paths },
-    );
-    // eslint-disable-next-line @typescript-eslint/no-require-imports
-    return require(packagePath);
-  } catch (e) {
-    return undefined;
-  }
+  // Adds the jsii runtime as a psuedo construct for reporting purposes.
+  allConstructInfos.push({
+    fqn: 'jsii-runtime.Runtime',
+    version: getJsiiAgentVersion(),
+  });
+
+  // Filter out duplicate values
+  const uniqKeys = new Set();
+  return allConstructInfos.filter(construct => {
+    const constructKey = `${construct.fqn}@${construct.version}`;
+    const isDuplicate = uniqKeys.has(constructKey);
+    uniqKeys.add(constructKey);
+    return !isDuplicate;
+  });
+}
+
+/**
+ * Returns all constructs under the parent construct (including the parent),
+ * stopping when it reaches a boundary of another stack (e.g., Stack, Stage, NestedStack).
+ */
+function constructsInStack(construct: IConstruct): IConstruct[] {
```

Comment:
Feels like this should go in `Stack`, no? Seems useful enough.

Reply:
Maybe. The fact that it's not there yet implies there may not be a second active use case. I'd prefer to refactor once we're written this code twice, rather than committing to a public API like `Stack` from this one use case.

Reply:
@rix0rrr isn't this required by stage/pipelines?